File: apps-microservices/crawler-monitor-backend/src/lib/domains.js
/**
 * Domain aggregation helpers.
 *
 * Each crawler job carries a `domain` field; we aggregate jobs by domain to
 * power the /domains list and the per-domain detail page (run chain).
 *
 * Window default: 7 days. Reuses the same loadJobs pattern as systemStats.
 */

const WINDOW_MAP = {
  '24h': 24 * 60 * 60 * 1000,
  '7d':  7 * 24 * 60 * 60 * 1000,
  '30d': 30 * 24 * 60 * 60 * 1000,
};

export function parseDomainWindow(input) {
  if (typeof input !== 'string' || !(input in WINDOW_MAP)) {
    throw new Error("Invalid window. Use '24h', '7d' or '30d'.");
  }
  return WINDOW_MAP[input];
}

const TERMINAL_OK = new Set(['finished', 'archived']);
const TERMINAL_KO = new Set(['failed']);

/**
 * Aggregate the list of jobs by domain.
 * Returns an array sorted by `last_run_at` desc.
 *
 * Each entry: {
 *   domain,
 *   total_jobs,
 *   success, failure, running,
 *   success_rate (0..1 or null if no terminal jobs),
 *   oom_total,
 *   update_share (0..1),
 *   last_run_at (ISO string or null),
 *   last_status,
 * }
 */
export function aggregateDomains(jobs, nowMs, windowMs) {
  const cutoff = nowMs - windowMs;
  const inWindow = jobs.filter(j => {
    const t = Date.parse(j.start_time);
    return Number.isFinite(t) && t >= cutoff && j.domain;
  });

  const byDomain = new Map();
  for (const j of inWindow) {
    const d = j.domain;
    let agg = byDomain.get(d);
    if (!agg) {
      agg = {
        domain: d,
        total_jobs: 0,
        success: 0, failure: 0, running: 0, other: 0,
        oom_total: 0,
        update_count: 0,
        last_run_at: null,
        last_run_ts: 0,
        last_status: null,
      };
      byDomain.set(d, agg);
    }
    agg.total_jobs++;
    const status = (j.status || '').toLowerCase();
    if (TERMINAL_OK.has(status)) agg.success++;
    else if (TERMINAL_KO.has(status)) agg.failure++;
    else if (status === 'running' || status === 'stopping' || status === 'restarting_oom') agg.running++;
    else agg.other++;
    agg.oom_total += j.oom_restart_count || 0;
    if (j.crawl_mode === 'update') agg.update_count++;
    const ts = Date.parse(j.start_time);
    if (Number.isFinite(ts) && ts > agg.last_run_ts) {
      agg.last_run_ts = ts;
      agg.last_run_at = j.start_time;
      agg.last_status = j.status || null;
    }
  }

  const out = [];
  for (const agg of byDomain.values()) {
    const completed = agg.success + agg.failure;
    out.push({
      domain: agg.domain,
      total_jobs: agg.total_jobs,
      success: agg.success,
      failure: agg.failure,
      running: agg.running,
      other: agg.other,
      success_rate: completed > 0 ? agg.success / completed : null,
      oom_total: agg.oom_total,
      update_share: agg.total_jobs > 0 ? agg.update_count / agg.total_jobs : 0,
      last_run_at: agg.last_run_at,
      last_status: agg.last_status,
    });
  }
  // Sort by latest activity desc
  out.sort((a, b) => Date.parse(b.last_run_at || 0) - Date.parse(a.last_run_at || 0));
  return out;
}

/**
 * Filter jobs for a single domain, sorted by start_time desc, with a built
 * "run chain" (links via previous_crawl_id) for the most recent root.
 */
export function jobsForDomain(jobs, domain, windowMs, nowMs = Date.now()) {
  const cutoff = nowMs - windowMs;
  const filtered = jobs
    .filter(j => j.domain === domain)
    .filter(j => {
      const t = Date.parse(j.start_time);
      return Number.isFinite(t) && t >= cutoff;
    })
    .sort((a, b) => Date.parse(b.start_time) - Date.parse(a.start_time));

  // Build a chain starting from the most recent job, walking back via previous_crawl_id.
  const byId = new Map(filtered.map(j => [j.id, j]));
  const chain = [];
  if (filtered.length > 0) {
    let cur = filtered[0];
    const seen = new Set();
    while (cur && !seen.has(cur.id)) {
      seen.add(cur.id);
      chain.push({
        id: cur.id,
        status: cur.status,
        start_time: cur.start_time,
        crawl_mode: cur.crawl_mode || null,
        oom_restart_count: cur.oom_restart_count || 0,
      });
      const prevId = cur.previous_crawl_id;
      cur = prevId ? byId.get(prevId) : null;
    }
  }

  return { jobs: filtered, chain };
}

File: apps-microservices/crawler-monitor-backend/internal/domain/domains/domains.go
// Package domains provides domain-level aggregation helpers.
//
// Each crawler job carries a `domain` field; we aggregate jobs by domain to
// power the /domains list and the per-domain detail page (run chain).
//
// Window default: 7 days. Reuses the same loadJobs pattern as systemStats.
package domains

import (
	"errors"
	"sort"
	"time"

	"github.com/Hellopro-fr/crawler-monitor-backend/internal/datetime"
)

var windowMap = map[string]int64{
	"24h": 24 * 60 * 60 * 1000,
	"7d":  7 * 24 * 60 * 60 * 1000,
	"30d": 30 * 24 * 60 * 60 * 1000,
}

// ParseDomainWindow parses "24h", "7d" or "30d" and returns duration in ms.
func ParseDomainWindow(input string) (int64, error) {
	ms, ok := windowMap[input]
	if !ok {
		return 0, errors.New("Invalid window. Use '24h', '7d' or '30d'.")
	}
	return ms, nil
}

var terminalOK = map[string]bool{"finished": true, "archived": true}
var terminalKO = map[string]bool{"failed": true}

// DomainSummary is an aggregated view of a domain's crawl activity.
type DomainSummary struct {
	Domain      string   `json:"domain"`
	TotalJobs   int      `json:"total_jobs"`
	Success     int      `json:"success"`
	Failure     int      `json:"failure"`
	Running     int      `json:"running"`
	Other       int      `json:"other"`
	SuccessRate *float64 `json:"success_rate"` // 0..1 or null
	OOMTotal    int      `json:"oom_total"`
	UpdateShare float64  `json:"update_share"`
	LastRunAt   *string  `json:"last_run_at"` // ISO string or null
	LastStatus  *string  `json:"last_status"`
}

// RawJob is the minimal job view consumed by domain aggregation.
type RawJob struct {
	ID              string `json:"id"`
	Domain          string `json:"domain"`
	StartTime       string `json:"start_time"`
	Status          string `json:"status"`
	CrawlMode       string `json:"crawl_mode"`
	OOMRestartCount int    `json:"oom_restart_count"`
	PreviousCrawlID string `json:"previous_crawl_id"`
}

// aggState tracks mutable state during domain aggregation (not exported).
type aggState struct {
	domain      string
	totalJobs   int
	success     int
	failure     int
	running     int
	other       int
	oomTotal    int
	updateCount int
	lastRunTs   int64
	lastRunAt   *string
	lastStatus  *string
}

// AggregateDomains groups jobs by domain and computes per-domain metrics.
// Returns an array sorted by last_run_at desc.
func AggregateDomains(jobs []RawJob, nowMs, windowMs int64) []DomainSummary {
	cutoff := nowMs - windowMs
	byDomain := make(map[string]*aggState)

	for i := range jobs {
		j := &jobs[i]
		if j.Domain == "" {
			continue
		}
		tMs := datetime.ParseStringMs(j.StartTime)
		if tMs < 0 {
			continue
		}
		if tMs < cutoff {
			continue
		}
		agg, exists := byDomain[j.Domain]
		if !exists {
			agg = &aggState{domain: j.Domain}
			byDomain[j.Domain] = agg
		}
		agg.totalJobs++
		status := j.Status
		if terminalOK[status] {
			agg.success++
		} else if terminalKO[status] {
			agg.failure++
		} else if status == "running" || status == "stopping" || status == "restarting_oom" {
			agg.running++
		} else {
			agg.other++
		}
		agg.oomTotal += j.OOMRestartCount
		if j.CrawlMode == "update" {
			agg.updateCount++
		}
		if tMs > agg.lastRunTs {
			agg.lastRunTs = tMs
			s := j.StartTime
			agg.lastRunAt = &s
			st := j.Status
			agg.lastStatus = &st
		}
	}

	out := make([]DomainSummary, 0, len(byDomain))
	for _, agg := range byDomain {
		completed := agg.success + agg.failure
		var successRate *float64
		if completed > 0 {
			sr := float64(agg.success) / float64(completed)
			successRate = &sr
		}
		updateShare := 0.0
		if agg.totalJobs > 0 {
			updateShare = float64(agg.updateCount) / float64(agg.totalJobs)
		}
		out = append(out, DomainSummary{
			Domain:      agg.domain,
			TotalJobs:   agg.totalJobs,
			Success:     agg.success,
			Failure:     agg.failure,
			Running:     agg.running,
			Other:       agg.other,
			SuccessRate: successRate,
			OOMTotal:    agg.oomTotal,
			UpdateShare: updateShare,
			LastRunAt:   agg.lastRunAt,
			LastStatus:  agg.lastStatus,
		})
	}
	// Sort by latest activity desc, matching JS: Date.parse(b.last_run_at||0) - Date.parse(a.last_run_at||0)
	sort.SliceStable(out, func(i, j int) bool {
		var ti, tj int64
		if out[i].LastRunAt != nil {
			ti = datetime.ParseStringMs(*out[i].LastRunAt)
		}
		if out[j].LastRunAt != nil {
			tj = datetime.ParseStringMs(*out[j].LastRunAt)
		}
		return ti > tj // desc
	})
	return out
}

// ChainEntry is a single entry in the run chain for a domain.
type ChainEntry struct {
	ID              string  `json:"id"`
	Status          string  `json:"status"`
	StartTime       string  `json:"start_time"`
	CrawlMode       *string `json:"crawl_mode"`
	OOMRestartCount int     `json:"oom_restart_count"`
}

// DomainDetail is the output of JobsForDomain.
type DomainDetail struct {
	Jobs  []RawJob     `json:"jobs"`
	Chain []ChainEntry `json:"chain"`
}

// JobsForDomain filters jobs for a single domain, sorted by start_time desc, with a
// "run chain" (linked via previous_crawl_id) starting from the most recent job.
func JobsForDomain(jobs []RawJob, domain string, windowMs, nowMs int64) DomainDetail {
	if nowMs == 0 {
		nowMs = time.Now().UnixMilli()
	}
	cutoff := nowMs - windowMs

	var filtered []RawJob
	for _, j := range jobs {
		if j.Domain != domain {
			continue
		}
		tMs := datetime.ParseStringMs(j.StartTime)
		if tMs < 0 || tMs < cutoff {
			continue
		}
		filtered = append(filtered, j)
	}
	// Sort newest first.
	sort.SliceStable(filtered, func(i, j int) bool {
		return datetime.ParseStringMs(filtered[i].StartTime) > datetime.ParseStringMs(filtered[j].StartTime)
	})

	// Build chain via previous_crawl_id starting from the most recent job.
	byID := make(map[string]*RawJob, len(filtered))
	for i := range filtered {
		byID[filtered[i].ID] = &filtered[i]
	}
	var chain []ChainEntry
	if len(filtered) > 0 {
		cur := &filtered[0]
		seen := make(map[string]bool)
		for cur != nil && !seen[cur.ID] {
			seen[cur.ID] = true
			var cm *string
			if cur.CrawlMode != "" {
				s := cur.CrawlMode
				cm = &s
			}
			chain = append(chain, ChainEntry{
				ID:              cur.ID,
				Status:          cur.Status,
				StartTime:       cur.StartTime,
				CrawlMode:       cm,
				OOMRestartCount: cur.OOMRestartCount,
			})
			if cur.PreviousCrawlID != "" {
				cur = byID[cur.PreviousCrawlID]
			} else {
				cur = nil
			}
		}
	}

	return DomainDetail{Jobs: filtered, Chain: chain}
}
